File: ABC/ABC301-400/ABC342/B/O(N+Q)/B.nim
import strutils, sequtils
let N = stdin.readLine.parseInt
let P = stdin.readLine.split().map(parseInt)
var Prev = newSeq[int](N)
for i in 0..<N:
  Prev[P[i] - 1] = i
let Q = stdin.readLine.parseInt
for i in 0..<Q:
  let a_b = stdin.readLine.split().map(parseInt)
  let
    A = a_b[0]
    B = a_b[1]
  echo if Prev[A - 1] < Prev[B - 1]: A else: B

File: ABC/ABC301-400/ABC342/B/O(N+Q)/B.py
N = int(input())
P = list(map(int, input().split()))
Prev = [0] * N
for i in range(N):
	Prev[P[i] - 1] = i
Q = int(input())
for _ in range(Q):
	A, B = map(int, input().split())
	print(A if Prev[A - 1] < Prev[B - 1] else B)